### PR TITLE
feat(prism-graph-example): add a hand-written provider for the contract

### PR DESCRIPTION
**Pull request by advl**
## Done

A second implementation of `@canonical/prism-contract`, written **by hand and
without the ke-graphql compiler**: a small metro-network dataset, a Relay
connection slicer, a total title chain, and a fetch handler. Private.

It is the control in a two-provider experiment. A documentation site that runs
against both this and the compiler-backed provider is provider-neutral by
*measurement* rather than by assertion — and this is the half that cannot be
neutral by accident, because nothing here shares a line with the compiler.

It lands **after** the site because its gate reads the site:
`src/testing/lensOperations.ts` discovers the site's lens operations from the
site's own tree at run time, out of the committed Relay artifacts, and executes
each against this provider.

Source: `feat/prism-docsite`, `24fe66cb0..6bbf65f9c`.

### Reconciled against `main`

Version `0.36.0` → `0.37.0`, biome `2.4.9` → `2.5.11` (one file reformats),
every `^0.36.0` sibling range → `^0.37.0`. The `biome.json` pinned `"$schema"`
at 2.4.9, which made `biome check` report a version mismatch on every run.

### Corrected while landing it

| What was wrong | Why it mattered |
| --- | --- |
| **`EntityMeta.title` was not total** | `lang` is an argument a client types, and the label lookup inherited `Object.prototype` — so `title(lang: "constructor")` returned a *function* where the schema promises `String!`. The field errored and nulled the whole node: the one outcome the fallback chain exists to make impossible |
| **The connection's page flags contradicted the Relay spec** | They took the cursor cut's answer where the spec takes the count's, so `first` with `before` reported a next page over a window that fit — a client following `endCursor` walks into an empty page. The two cells where the answers diverge were exactly the two the test matrix omitted |
| **The lens gate overstated its own scope** | Its prose said the scan root was the core set with exactly one documented exclusion, and that there was "no third state". There were five |
| **The green-null guard was weaker than documented** | It fired only when *every* root field came back `null`, so an empty list or an edgeless connection counted as having looked something up |
| **The barrels published the internals** | The dataset, the connection helpers, the SDL readers, eleven constants. The consumer boots this package and points at it; that is the surface now |
| **The operation scan collected `.tsx` only** | A `graphql` tag in one of the lens tree's `*Query.ts` modules would have been invisible to a gate whose whole claim is that the set comes from the tree |
| **`_meta.field(name:)` answered for unprojected properties** (second review pass) | `metro:note` is declared on the abstract `Stop` and inherited by every station, and no GraphQL type here exposes a `note` field. The provider handed back metadata for a field a consumer can never select — the contract says plainly that "there is no name `field(name:)` would accept" in that case. The lookup is gated on the built schema now: `fields` still publishes the local name as a label, `field(name: "note")` answers null |
| **The contract gate could not see an unimplemented contract field** (second review pass) | `readProviderSdl()` prepends the very contract `satisfiesContract` measures against, so a field added to the contract is in this schema the moment it is published, with nothing behind it. An execution gate now asks a real entity for every argument-free `EntityMeta` and `ClassProperty` field, with the selection built from the live schema so a new field is covered the day it lands |
| **The green-null guard stopped at the first non-empty array** (second review pass) | An empty connection inside a list item was never reached, so the "any connection anywhere" guarantee its own comment states was not the one the code kept |
| **The forward-compatibility shim was unreachable code** (second review pass) | It declared `ClassProperty.name` and `EntityMeta.curie` for a contract that lacked them; the supported range starts at 0.37.0, which has both. Removed, with its synthetic tests and its README section — 33 code lines |
| **The demo server bound every interface** (second review pass) | Bun binds `0.0.0.0` when `hostname` is omitted, and this process serves GraphiQL and permissive CORS outside production. It binds loopback now, as the sibling provider's demo does |
| **The README described the embeddable's `curie` backwards** (second review pass) | It said the value compacts its class (`geo:GeoPoint`); the code answers the GraphQL type name (`GeoPoint`), which is what the contract asks for |

## What this gate actually proves — please read this bit

The gate is green over **four of the documentation site's ten shipped query
operations**. The other six do not execute against a conformant provider at all:

| Where | Operations | Why |
| --- | --- | --- |
| `src/domains/components` | `ComponentsCatalogQuery`, `CatalogListPaginationQuery`, `ComponentEntityQuery` | select `Component` / `Query.component` |
| `src/domains/marketing` | `LobbyQuery` — **the home page** | selects `Component` |
| `src/domains/playground` | `ComponentProbeQuery` | selects `Query.component` |
| `src/addons/journeys` | `JourneysExplorerQuery` | selects `jobs` |

All six name pragma-ontology terms this provider's dataset has no class for.
Only the last was documented as an exclusion; the other five were invisible,
because the exclusion lived implicitly in one path constant while the prose
claimed there was "no third state".

So the honest claim is narrow: **the four operations under `src/domains/lenses`
are provider-neutral.** The site as a whole is not yet — five mounted routes,
the home page among them, cannot run against any conformant provider. Widening
the scan root to `src/domains` is the intended direction and would turn this
gate red today; that red is the to-do list, which is the state this gate was
built to be useful in. The exclusions are now enumerated in both the module
header and the README, with the reason for each.

## QA

Root gate, from the repository root:

- `bun run check` — 66 projects / 107 tasks, green
- `bun run test` — 49 projects, green; this package 10 files / 233 tests at 100% coverage
- `bun run build` — 55 projects, green
- `bun run check:ranges` — 66 workspace packages, every sibling range satisfied
- `bun run collect:check` — clean, `data/` unchanged

Environment notes, reproducible on unmodified `main`: the three
Playwright-tagged svelte projects cannot start a browser here
(`libglib-2.0.so.0` absent) and were excluded; the run used `--concurrency 1`.

## Carried

Raised in review, deliberately not fixed here:

- **The gate is invisible to `nx affected`.** The reach into the app is a
  filesystem path, not a manifest dependency, and the app depends on nothing
  here — so a PR that adds a lens does not schedule this project. Same shape as
  the provider package's gate, and the same trade: the alternative closes a
  cycle. Stated in the module header rather than left to be discovered.
- **Four dataset URIs are test-selection data exported from the data file**;
  they belong in `src/testing/fixtures.ts`.
- **`graphiqlHtml` and `localName` are noun-named** where
  `cs:code.naming.function_verb` wants verbs, and `descriptive.ts` /
  `connection.ts` are named for a topic rather than their primary export.
- **Bracket index access** in three places, where
  `@canonical/typescript-config` does not set `noUncheckedIndexedAccess`, so
  the type hides the `undefined`.
- **Two discovery regexes are tighter than the syntax they scan** — a
  variable-less query carrying a directive, and `@refetchable` with
  `queryName` other than first, are both skipped silently.
- **Several README measurements disagree with each other** (line counts for
  the same functions, reported two ways). The size table itself was recomputed
  when the shim came out, so its totals are current; the disagreement is
  between that table and figures quoted elsewhere in the prose.

## Readiness

- [x] Conventional-commit title
- [x] New behaviour covered by tests
- [x] Root gate green
- [x] Private package — no publish step

Chromatic: skip

https://claude.ai/code/session_01PUNAcTepWDfnyfNYqBnzbn

### CODE REVIEW

File: packages/prism/graph-example/src/lib/server/graphiqlHtml.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  </style>
+  <link rel="stylesheet" href="https://unpkg.com/graphiql@3.9.0/graphiql.min.css" />
+</head>
+<body>
+  <div id="graphiql">Loading GraphiQL…</div>
+  <script crossorigin src="https://unpkg.com/react@18.3.1/umd/react.production.min.js"></script>
+  <script crossorigin src="https://unpkg.com/react-dom@18.3.1/umd/react-dom.production.min.js"></script>
+  <script crossorigin src="https://unpkg.com/graphiql@3.9.0/graphiql.min.js"></script>
+  <script>
+    const fetcher = GraphiQL.createFetcher({ url: ${JSON.stringify(endpoint)} });
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/src/lib/server/graphiqlHtml.ts]: Correct, and the sibling template is the right precedent. `JSON.stringify` satisfies the JavaScript grammar, but the HTML tokenizer is what decides where the `<script>` element ends, and it decides by scanning the raw element text for `</script`, which JSON escaping leaves intact.

`graphiqlHtml.ts` now escapes every `<` through a local `asInlineScriptString` helper matching the one in `@canonical/ke-graphql`, and the comment says why it is every `<` rather than a `</script` match: the tokenizer has more than one exit (`<!--` moves it too) and a denylist would have to know them all.

A test feeds an endpoint carrying `</script><script>alert(1)</script>` and asserts both the escaped literal and that the document still holds exactly four closing script tags. Commit 72fc0992d.

File: packages/prism/graph-example/src/lib/provider/connection.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  // default this function supplies when it asked for neither. A defaulted
+  // `first` that did not trim says nothing about what lies beyond `before`,
+  // so in that case the cursor answer still stands — and a defaulted `first`
+  // that DID trim is itself proof of a next page.
+  const hasNextPage =
+    args.first !== undefined
+      ? trimmedByFirst
+      : trimmedByFirst || end < items.length;
+  const hasPreviousPage =
+    args.last !== undefined ? trimmedByLast : trimmedByLast || start > 0;
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/src/lib/provider/connection.ts]: All three stand, and they are three different bugs; thank you for separating them.

`sliceConnection` in `src/lib/provider/connection.ts` (commit 72fc0992d) now:

- throws on a negative `first` or `last` rather than clamping to zero, as the specification requires. Clamping made `first: -1` quietly return a default page, which is a client bug that never surfaces.
- computes BOTH page flags from the cursor-cut window, before either count trims it, which is how `HasNextPage` and `HasPreviousPage` are defined — "does ApplyCursorsToEdges hold more than `first` / `last`", not "did the cut fire". Your example is now a test: `first: 2, last: 4` over five items reports `hasPreviousPage: true`.
- treats an explicit `null` count as no count at all, so the cursor answer stands. Testing `!== undefined` had made `first: null` look like a client-supplied page size and answered from the default instead.

Three tests added, one per point, and the module header states each with the value the specification requires against the one the obvious implementation produces. Every existing case still holds unchanged.

File: packages/prism/graph-example/src/lib/provider/createExampleProvider.ts [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+        resolveTitle(source.labels, args.lang, source.uri, source.typename),
+      label: (args: LangArgs) => resolveLabel(source.labels, args.lang),
+      comment: (args: LangArgs) => resolveLabel(source.comments, args.lang),
+      definition: (args: LangArgs) =>
+        resolveLabel(source.definitions, args.lang),
+      // The compact display form of this entity's IRI. An embeddable has no
+      // IRI of its own, so it compacts its CLASS instead — still curie-shaped
+      // and still meaningful ("this is a geo:GeoPoint"), which a bare
+      // typename would not be.
+      curie: () => curieOf(source.uri ?? source.cls.uri),
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/src/lib/provider/createExampleProvider.ts]: You are right, and this is the provider that least ought to disagree with the contract. `EntityMeta.curie`'s third tier is spelled out at `contract.graphql:74-77` as the GraphQL type name for a value with no IRI at all, and compacting the class IRI answered `geo:GeoPoint` — which reads as an address the value does not have.

`metaView` now answers `source.typename` when `source.uri` is null, so an embeddable's `curie` is `GeoPoint`, matching the tail `_meta.title` already falls back to. The test that pinned the old answer is updated and quotes the contract's wording, so a future change back would have to argue with the contract rather than with a number.

`src/lib/provider/createExampleProvider.ts` and `createExampleProvider.test.ts`, commit 72fc0992d.

File: packages/prism/graph-example/src/lib/provider/createExampleProvider.ts [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+    servesLine: () =>
+      (entity.servesLine ?? []).flatMap((uri) => {
+        const line = viewByUri.get(uri);
+        return line === undefined ? [] : [line];
+      }),
+    servedBy: () =>
+      records
+        .filter((record) =>
+          (record.entity.servesLine ?? []).includes(entity.uri),
+        )
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/src/lib/provider/createExampleProvider.ts:370]: Correct. `servedBy` is the inverse of `servesLine`, `Interchange` carries `servesLine` too, and `[Station!]!` therefore described a list it did not hold: an interchange came back through it wearing the wrong type name, and `... on Interchange` could not be written against it at all.

Modelled as a shared interface, which is the option that keeps existing selections valid where a union would not (`servedBy { uri }` is already queried in the suite). `schema/extension.graphql` declares `interface Stop implements Node` over the fields both concrete types share, `Station` and `Interchange` both implement it, and `Line.servedBy` is `[Stop!]!`.

`Stop` is not an attempt to mirror the class tree — the hierarchy still lives in the TBox and surfaces through `superclasses` / `subclasses` / `inherited`. It is named after `metro:Stop`, the TBox class both types instantiate.

No resolver change was needed: the views already carry `__typename`, which is what graphql's default type resolver reads. A test walks `north-line`'s `servedBy` and asserts an `Interchange` comes back as one, with `transferMinutes` selectable through an inline fragment. Commit 72fc0992d.

File: packages/prism/graph-example/src/testing/integration/lensOperations.test.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+  }
+  /**
+   * Did a resolved root field actually look something up? An empty connection or
+   * an empty list is a green null wearing a different shape.
+   */
+  const isEmptyResult = (value: unknown): boolean => {
+    if (Array.isArray(value)) return value.length === 0;
+    if (typeof value === "object" && value !== null) {
+      const edges = (value as { edges?: unknown }).edges;
+      return Array.isArray(edges) && edges.length === 0;
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/src/testing/integration/lensOperations.test.ts]: Confirmed, and the operation that matters is exactly the one you name: `StandardsIndexQuery` selects `ontologyClass { instances { edges } }`, so the root field resolves to a perfectly good class and the empty connection sits two levels down, where a check reading one key off the root value never looked.

`isEmptyResult` now searches recursively: an empty array, or a connection anywhere beneath a root field whose `edges` are empty, counts as nothing looked up. The comment at the guard says why the connection is usually nested, so the next reader does not re-flatten it. `src/testing/integration/lensOperations.test.ts`, commit 72fc0992d.

File: packages/prism/graph-example/src/testing/integration/lensPagination.test.ts [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+  it("bonds every cursor to the identity it pages over", async () => {
+    // THE LOAD-BEARING ASSERTION OF THIS FILE. Both providers mint
+    // `base64(absolute IRI)`, which is the only reason a cursor from one is
+    // located by the other. This checks the derivation rather than its
+    // effect, because the effect of a broken cursor is indistinguishable
+    // from a first-page request.
+    const page = await fetchPage(null);
+
+    expect(page.edges.map((edge) => decodeCursor(edge.cursor))).toEqual(
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/src/testing/integration/lensPagination.test.ts:129]: The surrounding claim was overstated and the header is corrected. Every cursor this file reads is minted by this provider, so a change to ke-graphql's derivation would leave it green — it pins one side of the convention, not the agreement between the two.

What I have not done is make it cross-provider, and that is a constraint rather than a preference: obtaining a cursor from the compiler-backed provider means this package depending on `@canonical/ke-graphql`, which pins graphql v17 where this package peers v16. Two graphql majors in one process is the hazard the whole contract arrangement exists to keep out, and buying one assertion with it would be a bad trade.

So the two halves are pinned in two suites, and the header now says so and names the other one: ke-graphql's `src/testing/integration/queries.test.ts`, which asserts its cursors are base64 of the very IRI their node reports. This file is to be read as graph-example's half, not as a guard over both. `src/testing/integration/lensPagination.test.ts`, commit 72fc0992d.

File: packages/prism/graph-example/src/lib/provider/createExampleProvider.test.ts [hunk 7]
```diff
@@ -0,0 +1,1007 @@
...
+        // The key is `ClassProperty.name`, not the IRI. Only what `fields`
+        // publishes is accepted — that is what "round-trips" means.
+        notTheIri: null,
+      },
+    });
+  });
+
+  it("publishes a `name` on every field that `field(name:)` accepts back", async () => {
+    const data = await run(`{
+      node(id: "${MULTILINGUAL_ENTITY_URI}") { _meta { fields { name } } }
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/src/lib/provider/createExampleProvider.test.ts]: Stands, and this is the same finding as the one on `createExampleProvider.ts:233` — the test was pinning the behaviour that was wrong.

The round-trip now has two tests instead of one, because the contract makes two different promises. `publishes a name for every property, projected or not` keeps the full six-entry list including `note`, since `fields` should still publish it — the contract says the OWL local name is returned there as a label. `accepts every PROJECTED name back, and only those` round-trips the five names the schema actually declares as fields on this type, and then asserts `field(name: "note")` is `null`, with the contract's own sentence quoted as the reason.

A second test covers the other half of the gate, which the fix opened up: `namespace` IS a field on `OntologyClass`, and the metaclass declares no properties, so the schema check passes and there is still no `ClassProperty` to return. Commit `16ea5938f`.

File: packages/prism/graph-example/src/lib/provider/createExampleProvider.ts [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+      curie: () =>
+        source.uri === null ? source.typename : curieOf(source.uri),
+      type: () => classView(source.cls),
+      fields: () => fieldsOf(source.cls).map((field) => field.view),
+      // Matches `ClassProperty.name` exactly — the value this provider serves
+      // on every entry of `fields`. The argument round-trips: enumerate
+      // `_meta.fields`, read `name`, pass it straight back here.
+      field: (args: { name: string }) =>
+        fieldsOf(source.cls).find((field) => localName(field.uri) === args.name)
+          ?.view ?? null,
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/src/lib/provider/createExampleProvider.ts]: Stands. `metro:note` is declared on the abstract `Stop` and inherited by every station, and no type in `schema/extension.graphql` exposes a `note` field — so `field(name: "note")` was handing back cardinality metadata for a field a consumer can never select. The contract is explicit about that case: "there is no name `field(name:)` would accept; the OWL local name is returned as a label", and it adds that the fallback must never fabricate a plausible-but-wrong field name. This provider was fabricating one and then honouring it.

The lookup is now gated on the built schema rather than on a hand-kept list, so adding a field to the extension SDL is all it takes for that field to become addressable. `fieldNamesByType` is built once from `schema.getTypeMap()`, and `field(name:)` answers `null` unless the entity's own GraphQL type declares that field. `ClassProperty.name` is unchanged and still the property IRI's local name — for a projected property that string is the address, for an unprojected one it is the contract's label, and the comment at the field says so. There is deliberately no "unknown typename" arm: every typename reaching the check is a type this schema declares.

Tests as described on the sibling thread; the package is still at 100% branch coverage. Commit `16ea5938f`.

File: packages/prism/graph-example/README.md [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+   are in no declared namespace** and every curie would have degraded to the full IRI.
+   Serving `curie` well is therefore a constraint on how a provider *names its data*, not
+   just on how it answers a field. Instances moved into the declared namespaces, and a
+   `geo:Zone` class with real instances was added so that entities genuinely span two
+   prefixes — otherwise an implementation that hardcoded `metro:` would have passed every
+   test. The totality gate asserts that more than one prefix appears across the dataset.
+
+Edge cases, all gated: an IRI whose local name is empty compacts to a bare `metro:`; the
+embeddable, having no IRI of its own, compacts its **class** (`geo:GeoPoint`) rather than
+falling back to a bare typename, so the value stays curie-shaped.
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/README.md]: Stands — the README said the opposite of both the contract and the code.

`metaView` answers `source.typename` for a value with no IRI, which is the contract's own third tier quoted verbatim: "for a value with no IRI at all (an embedded blank node) — the GraphQL type name". So an embeddable answers `GeoPoint`, not `geo:GeoPoint`. The implementation comment already argued the point — compacting the CLASS reads better in isolation and is the wrong answer, and a provider whose job is to demonstrate the contract does not get to serve a nicer one — while the README described the nicer one as what happens.

The paragraph now states the typename tier, quotes the contract clause, and keeps the reason the class-compacting alternative was rejected. The "Naming note" beside it went with the shim it referred to (see the thread on `providerSdl.ts`). Commit `16ea5938f`.

File: packages/prism/graph-example/src/lib/provider/providerSdl.ts [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+/**
+ * Fields this provider serves that the contract is in the process of gaining,
+ * declared as `extend` only for as long as the contract lacks them.
+ *
+ * `ClassProperty.name` and `EntityMeta.curie` are landing in a parallel PR. A
+ * provider is free to be ahead of the contract — the superset rule allows any
+ * extra field — but it must not declare a field twice, which is what a static
+ * `extend` would do the moment the contract PR lands. So the decision is made
+ * from the contract text itself, and this whole function deletes cleanly once
+ * both fields are in.
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/src/lib/provider/providerSdl.ts]: Stands. `@canonical/prism-contract` is depended on at `^0.37.0`, and 0.37.0 declares both `ClassProperty.name` and `EntityMeta.curie` — so no supported installation can reach the branch that adds them, and the tests that exercised it were driving synthetic contract text through a code path production never takes.

`forwardCompatibleExtensions` is gone, with its `graphql` `parse` import and its six tests. `readProviderSdl` is now `[readContractSdl(), readExtensionSdl()].join("\n")` — the contract, then the extension, which is what the module's opening line always said it was. One test kept from that group, rephrased: the composed SDL builds.

The README's "A transitional shim, and its expiry date" section is deleted along with the "Naming note" that pointed at it, and the size table is recomputed rather than left with a subtraction row for something that no longer exists: `providerSdl.ts` is 33 code lines, and the provider runtime total is 420. Commit `16ea5938f`.

File: packages/prism/graph-example/src/testing/integration/lensOperations.test.ts [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+// deliberate and it is narrow. An add-on is a view that is openly
+// pragma-specific — it reads pragma's own ontology and makes no claim to run
+// anywhere else — and it is expected to become a plugin once the docsite has a
+// plugin mechanism. It is NOT a core lens, so it owes this gate nothing. It
+// remains fully built, routed and tested in the app.
+//
+// ONE operation lives there today: `JourneysExplorerQuery`, moved out of the
+// lens set by an OWNER DECISION, recorded verbatim: "keep it on side for now,
+// it will be an add-on plugin not a core view." It was reclassified because it
+// is not a core view; it was not reclassified because it was red.
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/src/testing/integration/lensOperations.test.ts]: Stands. The header presented a binary — core lens set here, add-ons there — and named only the journeys add-on as living outside the scan. Five of the six excluded operations are ordinary domain code, not add-ons, so a reader could reasonably take "everything except add-ons is covered" from it, which is not true.

The header now states the scope as four of ten, enumerates all six excluded operations with their directories, and separates the two reasons they are outside: five select `Component` or `Query.component`, pragma-ontology terms this provider's dataset has no class for, so widening the root to `src/domains` would turn the gate red today — a to-do list rather than a defect in the gate; the sixth is an add-on, excluded in kind rather than for failing. It closes with what the gate proves stated plainly: those four operations are provider-neutral, and the docsite as a whole is not, with five mounted routes including the home page currently unable to run against any conformant provider.

That is the same enumeration `lensOperations.ts` carries beside the path constant, and the header now says so, so the two cannot drift silently. Commit `16ea5938f`.

File: packages/prism/graph-example/demo/server.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+} from "../src/index.js";
+
+const port = Number(process.env.PORT ?? 5176);
+const provider = createExampleProvider();
+const handler = createExampleHandler(provider, {
+  endpoint: `http://127.0.0.1:${port}${GRAPHQL_PATH}`,
+});
+
+Bun.serve({
+  port,
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/demo/server.ts:34]: Stands. Bun binds every interface when `hostname` is omitted, and this process serves GraphiQL and permissive CORS outside production, so an unbound demo hands the local network a development GraphQL endpoint — one that also advertises itself as loopback in its own banner and GraphiQL endpoint, so the exposure is invisible from the thing you read when you start it.

`hostname: "127.0.0.1"` added, matching `packages/prism/pragma-provider/demo/server.ts`, with a comment saying what Bun does by default so the line is not deleted as redundant later. Commit `16ea5938f`.

File: packages/prism/graph-example/src/testing/integration/contract.test.ts [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+import { describe, expect, it } from "vitest";
+import {
+  createExampleProvider,
+  readProviderSdl,
+} from "../../lib/provider/index.js";
+
+describe("satisfiesContract(readProviderSdl())", () => {
+  // No `providerName`: `satisfiesContract` does not take one — it names the
+  // provider in a THROWN message, and this gate asserts on the result instead.
+  const result = satisfiesContract(readProviderSdl());
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/src/testing/integration/contract.test.ts:43]: Stands, and it is the more interesting of the two things you found in this file. `readProviderSdl()` prepends the very contract `satisfiesContract` uses as its baseline, so the schema half of this gate compares the contract with itself: it can only ever fail on the extension. A field added to the contract is in this provider's schema the moment it is published, with nothing behind it — nullable, it serves null and no test notices; non-null, it is a runtime error on a query nobody here runs.

Rather than narrow the claim, the gate gained the half it was missing. `contract.test.ts` now has an execution suite that builds its selection set FROM THE LIVE SCHEMA — every field of `EntityMeta` and of `ClassProperty` that can be selected without arguments, leaves selected bare and object-valued fields as `{ __typename }` so the resolver still has to run — asks a real entity for them, and asserts the execution reports no errors. A field taking a required argument (`field(name:)`) is skipped, since there is no value to invent for it, and it has its own tests.

Measured rather than assumed: adding a `probeAddedField: String!` to `EntityMeta` in the contract SDL leaves the subsumption tests green and turns the new one red with "Cannot return null for non-nullable field EntityMeta.probeAddedField" — which is your example, caught. The file header now explains why subsumption alone cannot see this, so the pairing is not undone by someone tidying it. Commit `16ea5938f`.

File: packages/prism/graph-example/src/testing/integration/lensOperations.test.ts [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+   * THE CONNECTION IS RARELY AT THE ROOT. `StandardsIndexQuery` selects
+   * `ontologyClass { instances { edges } }` — the root field resolves to a
+   * perfectly good class, and the connection that is empty is two levels down.
+   * Checking only the root value for an `edges` key therefore passed exactly
+   * the operations this guard exists for. So the search is RECURSIVE: any
+   * connection anywhere under a root field must have edges, or the field
+   * looked nothing up.
+   */
+  const isEmptyResult = (value: unknown): boolean => {
+    if (Array.isArray(value)) return value.length === 0;
```
[the remark this answers is not in the document]
Comment (advl) [reply packages/prism/graph-example/src/testing/integration/lensOperations.test.ts]: Stands. `if (Array.isArray(value)) return value.length === 0` answered "not empty" for any non-empty array and stopped there, so an empty connection sitting inside a list item — an edge's node, a class in a list of classes — was never reached. The comment directly above it promises "any connection anywhere under a root field", and that was not the guarantee the code kept.

It recurses through list members now: an array is empty either because it has no members or because one of its members is empty. The comment records the gap that was there, since the general shape of it (a search that stops one level short of what it claims) is easy to reintroduce.

The whole operation set still passes with the stricter search, so this tightens the guard without changing any verdict today. Commit `16ea5938f`.
